api/src/rpc.c

```diff
@@ -0,0 +1,369 @@
+    status = send_outgoing_rpc(p0, p1, p2, p3, gateway, 0, &cookie);
+    if (status) {
+        return status;
+    }
+
+    return cookie;
+}
+
+int rpc_fncall_wait(uvisor_rpc_result_t result, uint32_t timeout_ms, uint32_t * ret)
+{
+    int status;
+    uvisor_pool_slot_t const msg_slot = uvisor_result_slot(result);
+    uvisor_rpc_result_t const invalid = uvisor_result_build(UVISOR_RESULT_INVALID_COUNTER, msg_slot);
+
+    /* If the cookie is invalid, this message is already being waited for.
+     * Otherwise this is the first wait and we can proceed. */
+    uvisor_rpc_result_t cookie = __sync_val_compare_and_swap(
+            &outgoing_message_array()[msg_slot].wait_cookie,
+            result,
+            invalid);
+    if (cookie != result) {
+        return -1;
+    }
+
+    status = wait_for_rpc_result(msg_slot, timeout_ms);
+
+    if (status) {
+        return status;
+    }
+
+    /* The message result is valid now, because we woke up with a non-fatal
+     * status. */
+    *ret = outgoing_message_array()[msg_slot].result;
+
+    free_outgoing_msg(msg_slot);
+
+    return 0;
+}
+
+static uvisor_rpc_fn_group_t * allocate_function_group(const TFN_Ptr fn_ptr_array[], size_t fn_count)
+{
+    uvisor_pool_slot_t slot;
+    uvisor_rpc_fn_group_t * fn_group;
+
+    static const uint32_t timeout_ms = 0; /* Don't wait for an available slot. */
+    slot = uvisor_pool_queue_allocate(fn_group_queue(), timeout_ms);
+
+    if (slot >= fn_group_queue()->pool->num) {
+        /* Nothing left in pool */
+        return NULL;
+    }
+
+    fn_group = &fn_group_array()[slot];
+
+    fn_group->fn_ptr_array = fn_ptr_array;
+    fn_group->fn_count = fn_count;
+
+    uvisor_pool_queue_enqueue(fn_group_queue(), slot);
+
+    return fn_group;
+}
+
+/* Return a pointer to the function group specified by the fn_ptr_array, or
+ * NULL if that function group isn't found. */
+static uvisor_rpc_fn_group_t * get_function_group(const TFN_Ptr fn_ptr_array[], size_t fn_count)
+{
+    uvisor_pool_slot_t i;
+
+    /* Find the entry with linear search */
+    for (i = 0; i < fn_group_queue()->pool->num; i++) {
+        /* If found: */
+        if (fn_group_array()[i].fn_ptr_array == fn_ptr_array) {
+            /* Return the target. */
+            return &fn_group_array()[i];
+        }
+    }
+
+    /* The entry for the given function pointer wasn't found. */
+    return NULL;
+}
+
+/* Private structure for function group queries */
+typedef struct query_for_fn_group_context {
+    const TFN_Ptr *fn_ptr_array;
+    size_t fn_count;
+} query_for_fn_group_context_t;
+
+static int query_for_fn_group(uvisor_pool_slot_t slot, void * context)
+{
+    query_for_fn_group_context_t * query_context = context;
+    uvisor_rpc_message_t * msg = &incoming_message_array()[slot];
+    size_t i;
+
+    /* See if the target is for a function we can handle. */
+    for (i = 0; i < query_context->fn_count; ++i) {
```

Comment:
@Patater : Please add a FIXME with request for optimisation of this scan (hash table etc.).